apps/server/src/modules/tldraw/utils/utils.ts

```diff
@@ -0,0 +1,252 @@
+
+const CALLBACK_DEBOUNCE_WAIT = Configuration.get('FEATURE_TLDRAW_CALLBACK_DEBOUNCE_WAIT') as number ?? 2000;
+const CALLBACK_DEBOUNCE_MAX_WAIT = Configuration.get('FEATURE_TLDRAW_CALLBACK_DEBOUNCE_MAX_WAIT') as number ?? 10000;
+const pingTimeout = Configuration.get('FEATURE_TLDRAW_PING_TIMEOUT') as number ?? 30000;
+// disable gc when using snapshots!
+const gcEnabled = Configuration.get('FEATURE_TLDRAW_GC_ENABLED');
+
+enum wsStatesEnum {
+	wsReadyStateConnecting = 0,
+	wsReadyStateOpen = 1
+}
+
+enum wsMessageTypes {
+	messageSync = 0,
+	messageAwareness = 1
+}
+
+/**
+ * @type {{bindState: function(string,WSSharedDoc):void, writeState:function(string,WSSharedDoc):Promise<any>, provider: any}|null}
+ */
+let persistence: any = null;
+
+/**
+ * @param {{bindState: function(string,WSSharedDoc):void,
+ * writeState:function(string,WSSharedDoc):Promise<any>,provider:any}|null} persistence_
+ */
+export const setPersistence = persistence_ => {
+	persistence = persistence_;
+};
+
+/**
+ * @type {Map<string,WSSharedDoc>}
+ */
+const docs = new Map();
+
+/**
+ * @param {Uint8Array} update
+ * @param {any} origin
+ * @param {WSSharedDoc} doc
+ */
+const updateHandler = (update, origin, doc) => {
+	const encoder = encoding.createEncoder();
+	encoding.writeVarUint(encoder, wsMessageTypes.messageSync);
+	syncProtocol.writeUpdate(encoder, update);
+	const message = encoding.toUint8Array(encoder);
+	doc.conns.forEach((_, conn) => send(doc, conn, message));
+};
+
+export class WSSharedDoc extends Y.Doc {
+	private name: any;
+	conns: Map<string, any>;
+	private awareness: awarenessProtocol.Awareness;
+	/**
+	 * @param {string} name
+	 */
+	constructor(name) {
+		super({ gc: gcEnabled });
+		this.name = name;
+		this.conns = new Map();
+		this.awareness = new awarenessProtocol.Awareness(this);
+		this.awareness.setLocalState(null);
+
+		/**
+		 * @param {{ added: Array<number>, updated: Array<number>, removed: Array<number> }} changes
+		 * @param {Object | null} conn Origin is the connection that made the change
+		 */
+		const awarenessChangeHandler = ({ added, updated, removed }, conn) => {
+			const changedClients = added.concat(updated, removed);
+			if (conn !== null) {
+				const connControlledIDs = (this.conns.get(conn));
+				if (connControlledIDs !== undefined) {
+					added.forEach(clientID => {
+						connControlledIDs.add(clientID);
+					});
+					removed.forEach(clientID => {
+						connControlledIDs.delete(clientID);
+					});
+				}
+			}
+			// broadcast awareness update
+			const encoder = encoding.createEncoder();
+			encoding.writeVarUint(encoder, wsMessageTypes.messageAwareness);
+			encoding.writeVarUint8Array(encoder, awarenessProtocol.encodeAwarenessUpdate(this.awareness, changedClients));
+			const buff = encoding.toUint8Array(encoder);
+			this.conns.forEach((_, c) => {
+				send(this, c, buff);
+			});
+		};
+		this.awareness.on('update', awarenessChangeHandler);
+		this.on('update', updateHandler);
+		if (isCallbackSet) {
```

Comment:
Please move this `isCallbackSet` to this file as it's only used here (along with the code that sets its value from _Configuration_), preferably inside this class as its best to avoid globals wherever possible.